src/Storages/PostgreSQL/StoragePostgreSQLReplica.cpp

```diff
@@ -0,0 +1,514 @@
+    , replication_settings(std::move(replication_settings_))
+    , is_postgresql_replica_database(
+            DatabaseCatalog::instance().getDatabase(getStorageID().database_name)->getEngineName() == "PostgreSQLReplica")
+{
+    setInMemoryMetadata(storage_metadata);
+
+    auto metadata_path = DatabaseCatalog::instance().getDatabase(getStorageID().database_name)->getMetadataPath()
+                       +  "/.metadata_" + table_id_.database_name + "_" + table_id_.table_name;
+
+    replication_handler = std::make_unique<PostgreSQLReplicationHandler>(
+            remote_database_name,
+            connection_str,
+            metadata_path,
+            global_context,
+            replication_settings->postgresql_replica_max_block_size.changed
+                     ? replication_settings->postgresql_replica_max_block_size.value
+                     : global_context->getSettingsRef().max_insert_block_size.value,
+            replication_settings->postgresql_replica_allow_minimal_ddl.value, false);
+}
+
+
+StoragePostgreSQLReplica::StoragePostgreSQLReplica(
+    const StorageID & table_id_,
+    StoragePtr nested_storage_,
+    const Context & context_)
+    : IStorage(table_id_)
+    , global_context(std::make_shared<Context>(context_))
+    , nested_storage(nested_storage_)
+    , is_postgresql_replica_database(
+            DatabaseCatalog::instance().getDatabase(getStorageID().database_name)->getEngineName() == "PostgreSQLReplica")
+{
+
+}
+
+
+std::string StoragePostgreSQLReplica::getNestedTableName() const
+{
+    auto table_name = getStorageID().table_name;
+
+    if (!is_postgresql_replica_database)
+        table_name += NESTED_STORAGE_SUFFIX;
+
+    return table_name;
+}
+
+
+std::shared_ptr<ASTColumnDeclaration> StoragePostgreSQLReplica::getMaterializedColumnsDeclaration(
+        const String name, const String type, UInt64 default_value)
+{
+    auto column_declaration = std::make_shared<ASTColumnDeclaration>();
+
+    column_declaration->name = name;
+    column_declaration->type = makeASTFunction(type);
+
+    column_declaration->default_specifier = "MATERIALIZED";
+    column_declaration->default_expression = std::make_shared<ASTLiteral>(default_value);
+
+    column_declaration->children.emplace_back(column_declaration->type);
+    column_declaration->children.emplace_back(column_declaration->default_expression);
+
+    return column_declaration;
+}
+
+
+ASTPtr StoragePostgreSQLReplica::getColumnDeclaration(const DataTypePtr & data_type) const
+{
+    WhichDataType which(data_type);
+
+    if (which.isNullable())
+        return makeASTFunction("Nullable", getColumnDeclaration(typeid_cast<const DataTypeNullable *>(data_type.get())->getNestedType()));
+
+    if (which.isArray())
+        return makeASTFunction("Array", getColumnDeclaration(typeid_cast<const DataTypeArray *>(data_type.get())->getNestedType()));
+
+    /// getName() for decimal returns 'Decimal(precision, scale)', will get an error with it
```

Comment:
`Decimal(P, S)` should work fine in ClickHouse

Reply:
But I checked and it doesn't: `DB::Exception: Unknown data type family: Decimal(5, 5). Maybe you meant: ['Decimal256']
`

Reply:
Hm, looks like you found a bug in ClickHouse! Could you please provide a testcase for this? Unfortunately, I cannot reproduce it with trivial `CREATE`:
```
:) create table decimals (d1 Decimal(5, 5), d2 Decimal(25, 5)) engine=Memory

Ok.
```

Reply:
It is just [this test](https://github.com/ClickHouse/ClickHouse/pull/20470/files#diff-5708c43587e64584dc0d738979907a89b23ee2c6db30c32073c8f6a64b6878a4R162) from my tests. It works ok with my current code above, fails if I use `getName()` instead (like for other data types).